### ISSUE TITLE
sessionStateStream issue with Getx

### ISSUE THREAD

**Issue by FullstackWEB-developer** [SankethBK/local_session_timeout#26]
When I call this code `sessionStateStream.add(SessionState.startListening)` my app refreshes and redirects to the first page.
I followed example code, the only difference being that I used Getx package.
After spending a lot of time searching on Google, I created an issue to ask the package developers for help.

this is `main.dart` file in my project.
```
Future<void> main() async {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final sessionConfig = SessionConfig(
      invalidateSessionForAppLostFocus: const Duration(minutes: 5),
      invalidateSessionForUserInactivity: const Duration(minutes: 5),
    );
    sessionConfig.stream.listen((SessionTimeoutState timeoutEvent) {
      if (timeoutEvent == SessionTimeoutState.userInactivityTimeout) {
        // logout logic
      } else if (timeoutEvent == SessionTimeoutState.appFocusTimeout) {
        // logout logic
      }
    });
    return GetBuilder<SessionStateController>(builder: (sessionStateController) {
      return SessionTimeoutManager(
        sessionConfig: sessionConfig,
        sessionStateStream: sessionStateController.sessionStateStream.stream,
        child: GetMaterialApp(
          ...
          ),
        ),
      );
    });
  }
}
```

And this is `session_state_controller.dart` file
```
import 'dart:async';
import 'package:get/get.dart';
import 'package:local_session_timeout/local_session_timeout.dart';

class SessionStateController extends GetxController {
  final sessionStateStream = StreamController<SessionState>();

  void startListening() {
    if (!sessionStateStream.isClosed && sessionStateStream.hasListener) {
      sessionStateStream.add(SessionState.startListening);
    }
  }

  void stopListening() {
    if (!sessionStateStream.isClosed && sessionStateStream.hasListener) {
      sessionStateStream.add(SessionState.stopListening);
    }
  }

  @override
  void onClose() {
    sessionStateStream.close();
    super.onClose();
  }
}
```
I used this code when I click login button
```
    final sessionStateController = Get.put(SessionStateController());
    sessionStateController.startListening();
```

**Comment by SankethBK** [comment]
> When I call this code sessionStateStream.add(SessionState.startListening) my app refreshes and redirects to the first page.

Is it because of `userInactivityTimeout` or `appFocusTimeout`? you can print the `timeoutEvent` here
```
     if (timeoutEvent == SessionTimeoutState.userInactivityTimeout) {
        // logout logic
      } else if (timeoutEvent == SessionTimeoutState.appFocusTimeout) {
        // logout logic
      }
```

where are you calling `startListening` defined in `SessionStateController`?

**Comment by FullstackWEB-developer** [comment]
Hi @SankethBK 
Thank you for your reply.

> Is it because of userInactivityTimeout or appFocusTimeout? you can print the timeoutEvent here

No, it's only for the `sessionStateController.startListening();`.

> where are you calling startListening defined in SessionStateController?

After logging in to My Project, the dashboard page opens.
So I tried adding it to this part of the dashboard page, but the project structure is completely redrawn and it returns to the initial state when I started the app.
```
class _DashboardScreenState extends State<DashboardScreen> {
  @override
  void initState() {
    final sessionStateController = Get.put(SessionStateController());
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      sessionStateController.startListening();
    });

    SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
    ]);
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
     ...
  }
```

**Comment by FullstackWEB-developer** [comment]
![image](https://github.com/SankethBK/local_session_timeout/assets/92151436/764792f6-f73b-4515-952b-8dd1f659aa7a)
When I debug, it goes to this part and then redraws, so I don't think this is because of Getx.
I am new to this package, are there any other use cases using this package and Getx?

**Comment by SankethBK** [comment]
Hi @FullstackWEB-developer , ideally calling `setState` in the package should not cause rebuilds in child. But I think since you're using GetX for `MaterialApp`, the child widget is no longer treated as a const widget and rebuilds the entire widget tree

**Comment by SankethBK** [comment]
Can you try replacing `GetMaterialApp` with `MaterialApp` widget to confirm if this is the issue

**Comment by FullstackWEB-developer** [comment]
> Hi @FullstackWEB-developer , ideally calling setState in the package should not cause rebuilds in child. But I think since you're using GetX for MaterialApp, the child widget is no longer treated as a const widget and rebuilds the entire widget tree

So how can I fix this?
Shouldn't I use Getx?

> Can you try replacing GetMaterialApp with MaterialApp widget to confirm if this is the issue

Let me try but in my project I need to use `GetMaterialApp`

**Comment by SankethBK** [comment]
> So how can I fix this?
Shouldn't I use Getx?

I mean just to confirm if this is the bug. If using `MaterialApp` directly solves this issue, I'll raise the fix for this, instead of calling `setState` in my package, I'll handle it inside `_recordPointerEvents` function without calling `setState`

**Comment by FullstackWEB-developer** [comment]
If I don't use `GetMaterialApp` this isn't a problem.
But I am not sure how to integrate with Getx without using `GetMaterialApp`.

**Comment by SankethBK** [comment]
> If you don't use GetMaterialApp this isn't a problem for me.

Cool, I'll raise the fix in 1-2 hours you can test if it solves your issue

**Comment by FullstackWEB-developer** [comment]
Thanks

**Comment by SankethBK** [comment]
Try using this commit of the package `44d876f19667100e552a65ee672694939b7ed298`, I have removed the setState that fires immediately after calling `startListening`. If it works I'll remove all `setStates` they aren't necessary anyway

**Comment by FullstackWEB-developer** [comment]
Ok
Let me check soon

**Comment by SankethBK** [comment]
Fyi to reference package from github, add this in your pubspec.yaml

```
 local_session_timeout:
    git:
      url: https://github.com/SankethBK/local_session_timeout.git
      ref: 44d876f19667100e552a65ee672694939b7ed298 
```
and run `flutter pub get`

**Comment by FullstackWEB-developer** [comment]
Brilliant!
It's working well now.
Thank you for your help.

**Comment by SankethBK** [comment]
I assume other `setStates` might be still causing same issue, can you confirm once

**Comment by FullstackWEB-developer** [comment]
Yes
I checked it
![image](https://github.com/SankethBK/local_session_timeout/assets/92151436/c21a3111-c61b-4b03-9a36-c2e4c39d0d6e)
When debugging, these `setStates` were called, but re-rendering did not occur.
It's currently working well.
Nothing has changed on my end.

**Comment by SankethBK** [comment]
Great! can you check if it works with this commit as well `9bf30e872b3f8c1a7e662307a3ebbbf2f0598b86`, i have removed all `setStates` none of the expected functionality seems to break, if it works well in your case also I'll make a new release with these changes

**Comment by FullstackWEB-developer** [comment]
I tested it and it also working now

**Comment by SankethBK** [comment]
Cool, I have released the next version 2.3.2 with this bug fix
